ships/0031-shipwright-trigger.md

```diff
@@ -0,0 +1,383 @@
+
+As you can see on the diagram above, almost all components are interacting with the Inventory using the specialized query methods `SearchForGit` and `SearchForObjectRef`.
+
+#### Shipwright Build Controller
+
+The Builds are added or removed from the Inventory through the [Build Controller][triggerControllersBuildController], responsible to reflect all Shipwright Build resources into the Inventory. On adding new entries, the Build is prepared for the subsequent queries.
+
+### Tekton Run Controller
+
+Watches for [Tekton Run instances][triggerControllersRunController] referencing Shipwright Builds, when a new instance is created it creates a new `BuildRun`. The controller also watches over the `BuildRun` instance, in order to reflect the status back to the Tekton Run parent.
+
+The Tekton Run instances are part of the Custom Task workflow, everytime Tekton finds a `TaskRef` resource outside of Tekton's scope, it creates a `Run` instance with the coordinates. In other words, to extend Tekton's functionality third party applications must watch and interact with those objects.
+
+### Tekton PipelineRun Controller
+
+The [controller for PipelineRun instances][triggerControllersPipelineRunController] is meant to react when a Pipeline reaches the desired status, so upon changes on the resource the controller checks on the inventory if there are triggers configured for the specific resource in question, in the desired status.
+
+Upon the creation of a `BuildRun` instance, the `PipelineRun` object is labeled for the controller to be able to avoid reprocessing.
+
+### WebHook Handler
+
+The [WebHook handler][triggerWebHookHTTPHandler] is a simple HTTP server implementation which receives requests from the outside, and after processing the event, searches over Builds that should be activated. The search on the inventory happens in the same fashion as the controllers, however uses `SearchForGit` method.
+
+This type of `SearchForGit` is meant to match the repository URL, the type of event and the branches affected. For instance, the WebHook event can have different types, like `Push` or `PullRequest` and plus the branch affected.
+
+## Trigger Cause
+
+Whenever a `BuildRun` is issued by the Shipwright Triggers, it must register all the relevant attributes that caused it. The relevant attributes must be able to identify the which trigger inside the `.spec.trigger.when[]` slice, plus the context obtained from the WebHook request and from the Kubernetes resources watched.
+
+Those relevant attributes will be codified as `BuildRun` labels, following a predetermined structure on which the Build Controller is able to make use of. The Build Controller, the owner of `Build` and `BuildRun` resources, will then translate the information from the labels to properly structure status fields.
```

Comment:
I think these labels will also need to change the behavior of the BuildRun:

* when there is a trigger for a PullRequest, then the repository and revision (commit ID of the PR) will need to be passed to the source step.
* when there is a trigger for a Push, then the revision (commit ID that was pushed) will need to be passed to the source step.
* when there is a trigger for an Image, then I could imagine that the user will want this image to make it somewhere. This is not trivial because it depends on the strategy. For `ko`, I would want to do a [`goml set -f .ko.yaml -p defaultBaseImage -v THE_IMAGE_THAT_TRIGGERED`](https://github.com/herrjulz/goml#createupdate-properties-with-set). In a Dockerfile, I would want to pass it as a build-arg. So, we need maybe some strategy enhancements to support this. And we need to brainstorm how we can pass this.

While the Image part is probably for the future, we likely need the Git related things already at the beginning.

Reply:
interesting point @SaschaSchwarze0 ... I wonder if some further clarification is needed though as I read you text ^^.

I say that because  if one wants to change behavior, then generally in k8s you should use annotations for that vs. labels ... annotations are the more loosely coupled alternative to API change for controlling implementation behavior, where labels are metadata/useful information stored by the implementation when it is finished its intent,
like what @otaviof mentions in this section.  So your `labels will also need to change the behavior of the BuildRun` gave me some pause.

For example, the type of trigger matters, and the type of trigger might lead to different labels being set (git commit vs. image sha, etc.).  I could see that.

But then, if we have a "chain" of build runs that manifest from one build run leading to another build run getting triggered, that gets interesting.  Do we want shipwright interpreting labels from prior build runs in the chain and adjusting behavior?  Is that what you were thinking also?  Initially, that seems to not mesh with the k8s annotation vs. label convention.

But may I'm incorrectly reading in between the lines with what you intended.

Reply:
Hi @gabemontero, I was not so much looking at the label vs annotation aspect. But yeah, labels, also because of their length limitation could become a problem and annotations might be more correct - or some first-class fields of the BuildRun are used instead, but then it might be not obvious when they come from a user vs a controller (though not different to annotations that could also be user-defined).

Reply:
ok thanks for clarifying @SaschaSchwarze0 

also, good point on the label length restrictions, especially in the light on long'ish items like image or commit sha's; something the implementation would minimally have to be cognizant of 

I wonder @otaviof if the explicit field vs. label vs. annotation back and forth here should be captured as an open question.  I at least would be fine classifying it as such, merging this SHP with that, and then sorting this our during the implementation PR.  But of course let's see what sort of consensus arises as others chime in.  Perhaps a discussion topic for an upcoming community meeting.

Reply:
@SaschaSchwarze0, @gabemontero as we've spoken during the community meeting yesterday (#80) about the objective of passing information via labels, we have a couple o paragraph to summarize the objective with them. Please consider:

https://github.com/otaviof/community/blob/467cea4c797411ecfcc8e57facf7495b434cd15b/ships/0031-shipwright-trigger.md?plain=1#L297-L301

Let me know if that's specific and actionable as we expect.

Reply:
I like the latest text wrt the new trigger controller setting metadata that is then interpreted by the existing build controller, which then sets the first class fields in the status of the build run

The more I think about it, the more I am inclined that annotations and not labels should be set by the trigger controller to convey trigger cause, given the label length restrictions and the fact that we are "affecting" the existing controller's behavior and processing of the build run (in which status fields it sets), where k8s conventions say you should annotations for that.
Also, I would not think the existing build related controllers need to do queries based on any trigger labels, right?

But as I said before, I'm also fine with carrying the annotation vs. label decision as an open question to be sorted out during implementation PR review.  If you are not convinced to change from labels to annotations here @otaviof I at least would be good with just adding an open question on this, or if I'm the only vote for annotations and @SaschaSchwarze0 or others are good with labels.

I'm also more inclined to start with annotations vs. first class fields, simply to minimize the API impact.  We can always move from annotations to first class fields with relative ease.  The other direction is not as easy.

Assuming everyone else is fine with avoiding additional first class fields for this, I say we put to bed the annotation vs. label decision one way or the other and we can resolve this thread.

thanks

Reply:
> [...]
> The more I think about it, the more I am inclined that annotations and not labels should be set by the trigger controller to convey trigger cause, given the label length restrictions and the fact that we are "affecting" the existing controller's behavior and processing of the build run (in which status fields it sets), where k8s conventions say you should annotations for that. Also, I would not think the existing build related controllers need to do queries based on any trigger labels, right?
> 

That's a good point, Gabe! I'll make sure to update the EP to reflect this.

> I'm also more inclined to start with annotations vs. first class fields, simply to minimize the API impact. We can always move from annotations to first class fields with relative ease. The other direction is not as easy.

That's exactly the point, if needed be we can "promote" those fields later on.